Move layout related features to a separate app

warpech opened the pull request:
To implement https://github.com/Starcounter/RebelsLounge/issues/47, I have moved all the layout related features from Launcher to a separate app "LayoutProvider" in the same repo.

Consists of:

- `LauncherLayoutInfo` JSON namespace in merged partials (I cannot change the name of it, because it is hard-coded in `starcounter-include`
- Layout editor (press CTRL+E) to enable

The below screenshot shows Products + Images + LayoutProvider.

I have tried to avoid doing any improvements. I have focused only on splitting the code.

Pitfalls:

1. `/LauncherEditor/Master.html` is part of every merged partial.
2. I did not have a good idea *where* to show layout editor, once opened. I have decided to show it on top of the page and add a positive margin (300px) to `body` when the layout editor is opened.

@tomalec I think it is ready for review. @miyconst WDYT?

![2016-12-12 17_53_48-products](https://cloud.githubusercontent.com/assets/566463/21108043/f8594428-c093-11e6-9167-47e8200d6d63.png)

per-samuelsson commented:
> To implement Starcounter/RebelsLounge#47, I have moved all the layout related features from Launcher to a separate app "LayoutProvider" in the same repo.

I can't really give any stop or go as far as the pull request goes, but very generally, I think the idea to separate layout management and the launcher is great. 📯

per-samuelsson commented:
"LayoutProvider" => "AppLayouts"?

tomalec commented:
> "LayoutProvider" => "AppLayouts"?

I would rename it to omething compltely different.
That is Starcounter who provides layout features. Currently it does so by `<starcounter-include>` (end `juicy-composition` inside).
And the app devs could use layouts without any other app (Launcher not LayoutProvider) , even if the API is not enabled on server-side.
They could simply use default compositions.

The fact that it's not enabled by SC itself looks like a bug to me, and we should not make a app name based on that.

In long-term this app should be responsible for layout editing, management (and providing in that degree).

I'd rather name it like "CompositionsManager"

tomalec commented on src/Launcher/wwwroot/Launcher/viewmodels/LauncherTemplate.html:
```diff
@@ -35,31 +34,17 @@
                 <div id="launcher-menu">
                     <launcher-menu>
                         <ul>
-                            <template is="imported-template" content$="{{model.menu.LauncherLayoutInfo.MergedHtml}}" model="{{model.menu}}"></template>
```
Shouldn't this be a subject of separate PR?
Anyway, it is not valid HTML, same as 
```html
<ul>
  <div>
     <li>
     <li>
  </div>
</ul>
```

warpech replied:
As said in the OT, I did not make any changes here. 

You are right that this should be changed, but that is a different point in the roadmap (https://github.com/Starcounter/RebelsLounge/issues/53)

tomalec replied:
> As said in the OP, I did not make any changes here.

From what I can see at 
https://github.com/StarcounterPrefabs/Launcher/blame/87f9953ced77fe55b5f8bc25bf0416f13e2bd2e6/src/Launcher/wwwroot/Launcher/viewmodels/LauncherTemplate.html#L47

it was changed to `<starcounter-include>` by you.

> You are right that this should be changed,

What I'm trying to say is that: "It should not be changed in this PR, ans it should not be changed that way. "`<ul>` if needed should be a part of `starcounter-composition` or per-partial decorator.

warpech replied:
Oops, I agree you are right. A solution might be to add CompositionProvider to Starcounter and then this line can be reverted to use `imported-template`.

Per-partial decorator is not a solution at the moment, as we are too far from having it.

tomalec replied:
But why cant' we keep it as it was?

warpech replied:
Because `model.menu.LauncherLayoutInfo.MergedHtml` does not exist if LayoutProvider is not running.

warpech replied:
I think that this can only be fixed after CompositionProvider is enabled by default in Starcounter. Any better ideas?

warpech replied:
It is currently fixed using `imported-template` and a compute method in Polymer, but this fails in Firefox: https://github.com/StarcounterPrefabs/Launcher/pull/259#issuecomment-268817651

warpech commented:
> That is Starcounter who provides layout features. Currently it does so by `<starcounter-include>` (end juicy-composition inside). The fact that it's not enabled by SC itself looks like a bug to me, and we should not make a app name based on that.

I am not 100% convinced that actually making LayoutProvider as an app is such a terrible idea. It allows us to iterate quicker than Starcounter versions. Like in Android, many "system apps" are actually updated through the Play store.

> In long-term this app should be responsible for layout editing, management (and providing in that degree).

My suggestion then is to propose a pull request in Starcounter to include the `LauncherLayoutInfo` ([lines 28-103](https://github.com/StarcounterPrefabs/Launcher/blob/87f9953ced77fe55b5f8bc25bf0416f13e2bd2e6/src/LayoutProvider/Program.cs#L28-L103)). But before that happens we should be sure that it is proved to work in variuos cases.

Before that happens, I should just divide "LayoutProvider" into two apps?

1. CompositionProvider
2. CompositionManager

tomalec commented:
I'm not saying it's terrible idea. I think it's a great idea! :) but we could aim further for the future.

> Before that happens, I should just divide "LayoutProvider" into two apps?

Sounds great! If it is not too much work.

As then, everybody could use tiny "system app": "CompositionProvider"  to get "specific" server-side compositions available.
Then only thous who would like to use the GUI to manage those would run "CompositionManager"

tomalec commented:
> 1. `/LauncherEditor/Master.html` is part of every merged partial.
> 2. I did not have a good idea *where* to show layout editor, once opened. I have decided to show it on top of the page and add a positive margin (300px) to `body` when the layout editor is opened.

I think, WebSection idea, should solve those two.

But, why do we need to add this `Master.html` every where?

tomalec commented:
I would say that this is somehow messy, that layout app which is meant to help you handle compositions in clean way, will pollute all your compositions with additional slots.

warpech commented:
@tomalec I have pushed changes that fix resizing. Work from a bus, tested only in Chrome :)

tomalec commented on src/Launcher/bower.json:
```diff
@@ -12,17 +12,6 @@
-    "juicy-redirect": "Juicy/juicy-redirect#~0.4.0",
-    "juicy-color-picker": "Juicy/juicy-color-picker#gh-pages",
-    "iron-a11y-keys-behavior": "PolymerElements/iron-a11y-keys-behavior#~1.0.8",
-    "paper-icon-button": "PolymerElements/paper-icon-button#~1.0.5",
```
It's still being imported in `LauncherTemplate.html` and used in `style.css`

tomalec commented on src/Launcher/bower.json:
```diff
@@ -12,17 +12,6 @@
   ],
   "dependencies": {
-    "webcomponentsjs": "webcomponents/webcomponentsjs#~0.7.23",
-    "polymer": "Polymer/polymer#~1.7.0",
```
Launcher uses Polymer in many places, so I'd keep this dependency

warpech replied:
Fixed

tomalec commented on src/LayoutProvider/bower.json:
```diff
@@ -0,0 +1,19 @@
+    "tests"
+  ],
+  "dependencies": {
+    "paper-icon-button": "PolymerElements/paper-icon-button#~1.0.5",
```
I would add Polymer as well, as we use just `dom-bind`s

warpech replied:
Fixed

tomalec commented on src/LayoutProvider/Program.cs:
```diff
@@ -0,0 +1,105 @@
+                returnNewSibling = true;
+                StarcounterEnvironment.RunWithinApplication(layoutNamespace, () => {
+                    layoutInfo = new LayoutInfo();
+                    layoutInfo.Html = "/LayoutProvider/Master.html";
```
I would say that this is somehow messy, that layout app which is meant to help you handle compositions in clean way, will pollute all your compositions with additional slots.

Plus, as you merge into every partial, the compostiiotns that will be prepared will be saved for `.*LayoutProvider=Master.html`. Therefore, once you will stop the app, all will no longer apply to any partial. 

I believe, it's a key feature of app to be able to:

1. Run the app in dev environment
2. Edit & save HTMLCompositions using GUI tools,
3. 
   1. Stop the app and use environemtn as production one or 
   2. Migrate the compositions to production environments

warpech replied:
I don't think that this is messy, this is just a paradox of having all of the apps adhere to the same principle.

Let's think of possible solutions:

1. Overwrite `starcounter.html` static file to add layout editor capabilities. This is short-sighted, because more apps might attempt to do that and only one can win
2. Extend `starcounter.html` in middleware. This is slightly better. Unfortunately, because it is a static file, it will not give a chance to write server side logic to it. Maybe it can still work in hybrid - have a `LayoutProvider` namespace at every level in the merged JSON, but without HTML property.
3. LayoutProvider could wrap the entire response in its own `PublicViewModel` just as Launcher does. I wonder what happens if one app wants to wrap` PublicViewModel` that already wraps another `PublicViewModel`.

@tomalec are you skewed towards any of the proposed solutions, or do you have some other?

tomalec replied:
4. 
   1. Return editor HTML - button that toggles it - only for `/CompositionManager/editor/` CompositionEditorTrigger :: WebSection 
   2. Map it to the `/launcher/user` WebSection

Thanks to that the page will look and exactly as before, we will be able to achieve start->edit->stop behavior I described at https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r93133820
No hack or big rewrites will be needed, we will do it Starcounter-way: micro-app that does small thing mapped to bigger suite.

We could discuss the possibilities for merged partials for mapped apps w/o HTML Partial, as separate and orthogonal issue.
Same goes for mapping to root node.

warpech commented on src/LayoutProvider/wwwroot/LayoutProvider/Master.html:
```diff
@@ -0,0 +1,58 @@
+            <paper-icon-button id="LayoutProvider-icon" active="{{model.editingMode$}}" src="/LayoutProvider/images/layout-editor.svg" toggles></paper-icon-button>
+        </template>
+        <template is="dom-if" restamp if="{{model.layoutModified$}}">
+            <paper-icon-button id="LayoutProvider-icon" active="{{model.editingMode$}}" src="/LayoutProvider/images/layout-editor-unsaved.svg" toggles class="modified" title="You have unsaved layout changes on this page!"></paper-icon-button>
```
First of all, I would rather not do performance optimisations at this point. 

There is a bigger problem: `layoutModified$` is always false, because `starcounter-layout-html-editor` does not inform the view-model about the dirty state. Can we fix this as an issue after this PR?

warpech commented on src/LayoutProvider/wwwroot/LayoutProvider/Master.html:
```diff
@@ -0,0 +1,58 @@
+        </template>
+
+        <layoutprovider-hot-key keys="Ctrl+E" on-keypress="toggleEditor"></layoutprovider-hot-key>
+        <layoutprovider-hot-key keys="Esc" on-keypress="closeEditor"></layoutprovider-hot-key>
```
Fixed

warpech commented on src/LayoutProvider/wwwroot/LayoutProvider/Master.html:
```diff
@@ -0,0 +1,58 @@
+        <layoutprovider-hot-key keys="Esc" on-keypress="closeEditor"></layoutprovider-hot-key>
+
+        <template is="dom-if" restamp if="{{model.editingMode$}}">
+            <template class="layout-editor-inserter" is="imported-template" model="{{model}}" content="/LayoutProvider/LayoutEditor.html">
```
Fixed by switching to `<layoutprovider-editor-inserter>`

warpech commented:
@tomalec, thanks a lot for the review. I have fixed most of the problems.

Still to be done in this PR:

- `<starcounter-include>` breaks `<ul>` in the menu (see my comment in https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r92128784)
- do not add `/LayoutProvider/Master.html` to the merged partial URL. I totally agree that this is needed. Please respond to my comment in https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r93133820.
- split LayoutProvider into CompositionProvider and CompositionManager. I will do it as the last step, because it will mess up all the comments.

Should be separate issues after this PR: 

- remove styles for `paper-icon-button` and `so-signin` (https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r93274586)
- optimize how paper-icon-button is inserted in Master.html (https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r92158958)

warpech commented:
@tomalec can you pls re-review my changes from today? 

Despite our conversation yesterday, I have searched for a solution that works *now* and decided to propose using middleware to extend `starcounter.html`.

The above commit makes the layout editor work without adding `Master.html` to the partial URI. In details:

- it uses Middleware to attach `layoutprovider-editor-inserter` to `/sys/starcounter.html` response
- `layoutprovider-editor-inserter` is aware that it can be used from both hot key and a UI button
- the hot key is CTRL+E to toggle and Esc to close
- UI button is mapped by default to Launcher's "user" area

Because I extended static file, layout editor needs to keep communicating only through DOM and the REST API. This will be a problem when the layout editor needs to have more server side logic. Implementing multiple JSON roots is a solution to that (https://github.com/Starcounter/Starcounter/issues/4017).

Still need solving:

- `<starcounter-include>` breaks `<ul>` in the menu (see my comment in https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r92128784). This can be solved when we have auto-start apps (https://github.com/Starcounter/CompanyTrack/issues/259). I will check the status of that with @alemoi
- split LayoutProvider into CompositionProvider and CompositionManager. I will do it as the last step, because it will mess up all the comments.

warpech commented:
The above commit fixes:

- - `<starcounter-include>` breaks `<ul>` in the menu (see my comment in https://github.com/StarcounterPrefabs/Launcher/pull/259#discussion_r92128784)

warpech commented:
The above commit does the last remaining step:

- split LayoutProvider into CompositionProvider and CompositionEditor. I will do it as the last step, because it will mess up all the comments.

warpech commented:
@tomalec please go with the final review. Meanwhile I will start discussing how to include CompositionProvider in Starcounter with @alemoi.

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html:
```diff
@@ -0,0 +1,74 @@
+                }
+            },
+            detached: function() {
+                this.active = false;
```
Shouldn't it detach panel?

warpech replied:
It does, because `this.active = false` has an observer at line 26.

tomalec replied:
ah.. I thought it requires `this.set('active', false)`; to notify observers

warpech replied:
I did this without thinking, but apparently it works.

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html:
```diff
@@ -0,0 +1,74 @@
+</dom-module>
+<script>
+    (function() {
+        function runIfNotExist(id, callback) {
```
I think HTML Import would be enough. Imported document is parsed only once, no matter how many time you import it.

warpech replied:
Good point, I will fix.

warpech replied:
Done

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-panel/compositioneditor-panel.html:
```diff
@@ -0,0 +1,34 @@
+-->
+<script src="/sys/ace-builds/src-noconflict/mode-html.js"></script>
+
+<dom-module id="compositioneditor-panel">
```
For performance reasons I'd make it Vanilla element, however maybe we can keep it to get Polymer's hybrid mode for migration from v0 to v1. (Need to check if SD v01 v1 polyfills work together).

It would be nice to add it as a code comment.

warpech replied:
I would really like to finish this PR this year... ;)

tomalec replied:
:) then just leave it as a code comment, and we will process it later

warpech replied:
Done

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-hot-key/compositioneditor-hot-key.html:
```diff
@@ -10,23 +10,27 @@
                 var that = this;
 
-                document.addEventListener("keydown", function (e) {
+                this.listener = function(e) {
```
I'd move it to `created`, not to redefine it when re-attached

warpech replied:
So you say that if I do it in `created`, I don't need to call `removeEventListener` in `detached`? It will detach automatically when the element is removed from DOM?

warpech replied:
I have just tested an unfortunately removing the element from DOM in Chrome Dev Tools does not automatically cancel the listeners. So I see no point in changing this to `created`.

tomalec replied:
> So you say that if I do it in created, I don't need to call removeEventListener in detached? It will detach automatically when the element is removed from DOM?

No, you need to call `removeEventListener`.

I'm saying that if you 
1. create `compositioneditor-hot-key`
2. attach it
3.  detach it
4. attach it once again 
5. ...repeat 2-4

the execution will be as following

1.   
2. Create `this.listener`,  `document.addEventListener(..)`
3. `removeEventListener(..)`
4. Create `this.listener`,  `document.addEventListener(..)`
5. ...repeat 2-4

What I propose is:
1. Create `this.listener`, 
2. `document.addEventListener(..)`
3. `removeEventListener(..)`
4. `document.addEventListener(..)`
5. ...repeat 2-4


Can you now see the "point in changing this to `created`"?

warpech replied:
So you mean just specifying:

```js
created: function() {
  this.listener = null;
}
```

and live the rest as is?

tomalec replied:
I mean 
```js
created: function(){
    var that = this;
    this.listener = function(e) {
        var keys = that.keys.split(" ");
        for (var i = 0; i < keys.length; i++) {
            if (that.isMatch(e, keys[i])) {
                e.preventDefault();
                that.fire("hotkeydown", e);
                return;
            }
        }
    };
},
attached: function () {
    document.addEventListener("keydown", this.listener);
},
```
and leave the rest as is?

warpech replied:
Ah sure, done!

tomalec commented on src/CompositionEditor/Program.cs:
```diff
@@ -0,0 +1,39 @@
+            {
+                if (request.Uri == "/sys/starcounter.html")
+                {
+                    var inserter = @"<link rel=""import"" href=""/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html"">";
```
I would add a code comment, on why do we need such kind of magic/hacking

warpech replied:
Done

tomalec commented:
Is it supposed to run with SC 2.3?
I got the following error when I try to run Launcher (Editor or Composer) 
```log
ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.
Version: 2.3.0.5406.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10007.

Type "star -h" to see help
Type "star 10007" to launch the help page for error 10007

Stack trace:
   w Starcounter.Server.ServerServiceProcess.WaitUntilServerIsOnlineOrSignalExit[T](Func`2 checkExited, T context) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\ServerServiceProcess.cs:wiersz 210
   w Starcounter.CLI.StartServerCLICommand.RunWithinContext() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.CLI\StartServerCLICommand.cs:wiersz 53
   w Starcounter.CLI.StartApplicationCLICommand.Run() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.CLI\StartApplicationCLICommand.cs:wiersz 120
[12:05:57, Starcounter (service)]
Process exitcode: 4259

[12:05:57, Starcounter (service)]
ScErrDatabaseEngineTerminated (SCERR10007): The server detected that the user code host executable (sccode.exe) unexpectedly terminated during a management operation.

[12:05:57, Starcounter (administrator)]
Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. Starcounter.Advanced.Configuration.InvalidConfigurationException: Cannot retrive database uuid. Database name: DEFAULT, path:
C:\Users\       omalec\Documents\Starcounter\Personal\Data\Default\Default-20161216T134654029
   w Starcounter.Server.Database.ReadDbUUID(DatabaseConfiguration configuration) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\Database.cs:wiersz 184
   w Starcounter.Server.Database..ctor(ServerEngine server, DatabaseConfiguration configuration) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\Database.cs:wiersz 124
   w Starcounter.Server.ServerEngine.SetupDatabases() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\ServerEngine.cs:wiersz 251
   w Starcounter.Server.ServerEngine.Setup() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\ServerEngine.cs:wiersz 205
   w Starcounter.Administrator.Server.Program.Main(String[] args) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Starcounter.Administrator\Server\Program.cs:wiersz 62
Version: 2.3.0.5406.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4259.
   w Starcounter.Hosting.Package.ExecuteEntryPoint(Application application, Assembly assembly) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Package.cs:wiersz 542
   w Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application, ApplicationDirectory applicationDir) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Package.cs:wiersz 265
   w Starcounter.Hosting.Package.Process() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Package.cs:wiersz 138
   w Starcounter.Hosting.Package.Process(IntPtr hPackage) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Package.cs:wiersz 55
   w StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:wiersz 85
ErrorCode=4259
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4259
---> System.Reflection.TargetInvocationException: Obiekt docelowy wywolania zglosil wyjatek.
   w System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments,
Signature sig, Boolean constructor)
   w System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   w System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   w Starcounter.Hosting.Package.ExecuteEntryPoint(Application application, Assembly assembly) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Hosting\Starcounter.Hosting\Package.cs:wiersz 529
HResult=-2146232828
---> Starcounter.Advanced.Configuration.InvalidConfigurationException: Cannot retrive database uuid. Database name: DEFAULT, path: C:\Users\    omalec\Documents\Starcounter\Personal\Data\Default\Default-20161216T134654029
   w Starcounter.Server.Database.ReadDbUUID(DatabaseConfiguration configuration) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\Database.cs:wiersz 184
   w Starcounter.Server.Database..ctor(ServerEngine server, DatabaseConfiguration configuration) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\Database.cs:wiersz 124
   w Starcounter.Server.ServerEngine.SetupDatabases() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\ServerEngine.cs:wiersz 251
   w Starcounter.Server.ServerEngine.Setup() w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Server\Starcounter.Server\ServerEngine.cs:wiersz 205
   w Starcounter.Administrator.Server.Program.Main(String[] args) w C:\TeamCity\BuildAgent\work\sc-669\Level1\src\Starcounter.Administrator\Server\Program.cs:wiersz 62
HResult=-2146232832

```

tomalec commented:
The last thing I found is that brush icon in user region does not look as pretty as before.
(size and background)
![fireshot capture 008 - products - http___localhost_8080_products_new-product](https://cloud.githubusercontent.com/assets/17435/21424455/54684596-c843-11e6-9684-36dce648a22d.png)

warpech commented:
> The last thing I found is that brush icon in user region does not look as pretty as before.
(size and background)

This is a problem, because in the past the icon was white on transparent background. Imagine what happens when someone shows the icon on white background. Suggestions?

tomalec commented:
That would be a common problem (luckily we have blending;) )
I think most natural would be to assume black on transparent, then maybe Launcher should style it otherwise

warpech commented:
@tomalec I have just noticed that Launcher menu does not work in Firefox with the error:

```
[dom-bind::_annotatedComputationEffect]: compute method `getMenuHtml` not defined
```

Any ideas?

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html:
```diff
@@ -0,0 +1,66 @@
+        insertHotKey("CompositionEditor/hot-key-close", "Esc", function() {
+            var panel = document.querySelector("compositioneditor-inserter");
+            if (panel) {
+                if (isPanelMine) {
```
What is it for?

warpech replied:
If the panel was opened by hot key, detach the panel completely. If someone else has inserted the panel (`Master.html`), be more polite and toggle the `active` property.

The purpose of this is to have cleaner DOM when `Master.html` is not attached in user area.

The alternative is to always include `compositioneditor-inserter` on page load, but without the active attribute.

tomalec replied:
> detach the panel completely
But you are not selecting the panel, but just inserter. And you are detaching inserter. Sorry, I'm quite lost, on what you call a "panel".


What if there are two `compositioneditor-inserter`s?
I'd rather do it that way, that all instances of `compositioneditor-inserter` shares the same `compositioneditor-panel`.

Then The way I'd use it is 
- `Master.html` stamps `<compositioneditor-inserter>` and binds it to icon-button
- `compositioneditor-inserter.html` stamps `<compositioneditor-hot-key>` and attaches the callback for them.
  - when open hot key is called it attaches its `<compositioneditor-inserter>` to the body (if not there), and activates it
  - when close hot key is called it de-activates its `<compositioneditor-inserter>` (which should reflect in `activate==false` in all other instances)
  - when open hot key is called again it activates it (and reflects in other)

So it's more-or-less, lazy-loaded: "The alternative is to always include compositioneditor-inserter on page load, but without the active attribute."
+ shared panel and active state

warpech replied:
Please check my last commit, now this more or less implement this idea.

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html:
```diff
@@ -0,0 +1,66 @@
+        var isPanelMine = false;
+
+        insertHotKey("CompositionEditor/hot-key-toggle", "Ctrl+E", function() {
+            var panel = document.querySelector("compositioneditor-inserter");
```
what if there are more inserters?

tomalec commented:
>  I have just noticed that Launcher menu does not work in Firefox with the error:
 ```
[dom-bind::_annotatedComputationEffect]: compute method `getMenuHtml` not defined
  ```
> Any ideas?

I'm afraid that may be side-effect of our workaround for a side-effect of a workaround for Polymer/polyfill/Chrome bug :(

tomalec commented:
Maybe then we should make a launcher-menu custom element, which will provide `getMenuHtml` and `<ul>`?

tomalec commented:
And wait for
```html
<ul>
    <template is="starcounter-include-decorated" partial="{{model.menu}}">
         <li><slot></slot></li>
     </template>
</ul>
```

tomalec commented:
2. Or maybe we should do
```html
<starcounter-include partial="{{model.menu}}">
    <template is="starcounter--partial-decorator">
        <ul class="launcher-menu"><slot></slot></ul>
    </template>
</starcounter-include>
```

warpech commented:
Or... 3. Revert to 

```html
<template is="imported-template" 
content$="{{model.menu.LauncherLayoutInfo.MergedHtml}}" 
model="{{model.menu}}"></template>
``` 

In case `model.menu.LauncherLayoutInfo` is undefined display a message: `Launcher menu requires HTML Composition to be on. *Click here* for more info.` And then a link to an issue where we work out a long term solution.

tomalec commented:
> 3. In case model.menu.LauncherLayoutInfo is undefined display a message: Launcher menu requires HTML Composition to be on. *Click here* for more info. And then a link to an issue where we work out a long term solution.

This implies tight dependency between specific apps.

tomalec commented:
4. `starcounter.html` may expose global `Starcounter.utils.buildMergedURL(partial)`

warpech commented:
> 4. `starcounter.html` may expose global `Starcounter.utils.buildMergedURL(partial)`

But how to use it in `LauncherTemplate.html` if compute methods do not work?

tomalec commented:
They do work. There is just a race between dom-bind which needs them, and polyfill/our workaround that executes the following script which defines it. If the method will be global it should work in FF.

warpech commented:
5. A 2-liner that puts <starcounter-include> as a sibling to <ul>. This works because <launcher-menu> doesn't care:

![2016-12-22 23_29_02-sourcetree](https://cloud.githubusercontent.com/assets/566463/21442534/a96b19fa-c89e-11e6-9748-6a5c62304663.png)

warpech commented:
> They do work. There is just a race between dom-bind which needs them, and polyfill/our workaround that executes the following script which defines it. If the method will be global it should work in FF.

I tried this but I cannot get it to work. Can you make a PR to this PR to show it? Sounds like a good idea.

tomalec commented:
> A 2-liner that puts as a sibling to

   - . This works because doesn't care:

Does it work because you simply disregard ul and displays just block
elements?

22 gru 2016 23:36 "Marcin Warpechowski" <notifications@github.com>
napisał(a):

> They do work. There is just a race between dom-bind which needs them, and
> polyfill/our workaround that executes the following script which defines
> it. If the method will be global it should work in FF.
>
> I tried this but I cannot get it to work. Can you make a PR to this PR to
> show it? Sounds like a good idea.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/StarcounterPrefabs/Launcher/pull/259#issuecomment-268906057>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABEG76I-gjBQ5LcGauAwHW-oICJYBacks5rKvtmgaJpZM4LKxaJ>
> .
>

warpech commented:
> Does it work because you simply disregard ul and displays just block
elements?

No, it works because `<launcher-menu>` removes all observed `<li>` elements from local DOM, sorts them by text content and appends them to the first found`<ul>` in local DOM.

It doesn't care if the `<li>`s are found in `<starcounter-include>`, which is a child or a sibling of `<ul>`.

tomalec commented on src/CompositionEditor/wwwroot/CompositionEditor/elements/compositioneditor-inserter/compositioneditor-inserter.html:
```diff
@@ -0,0 +1,66 @@
+        });
+
+        insertHotKey("CompositionEditor/hot-key-close", "Esc", function() {
+            var panel = document.querySelector("compositioneditor-inserter");
```
I'd rename it from `panel` to `inserter` not to confuse with `<compositioneditor-panel>`

warpech replied:
Good catch, I was playing with it and forgot to change accordingly.

warpech replied:
Fixed by removal :)

warpech commented:
@tomalec any can we proceed with merging? It is important to let @alemoi continue with including CompositionProvider in Starcounter. I propose the following steps.

1. Release Launcher 5.0.2, the final version before layouts stuff is moved out from Launcher
2. Merge this PR
3. Move CompositionProvider and CompositionEditor to separate repos (in Starcounter org)
4. Release Launcher 6.0.0, CompositionProvider 1.0.0, CompositionEditor 1.0.0
5. Ask @alemoi to proceed with including CompositionProvider in Starcounter (https://github.com/Starcounter/Starcounter/issues/4019)

tomalec commented:
Composition provider repo: https://github.com/Starcounter/CompositionEditor

warpech commented:
Thank you @tomalec.

There is no sense in putting Launcher 5.0.2 with dependency at Starcounter 2.3.x to the old warehouse, because Starcounter 2.3.x uses the new warehouse. See point 8 in this thread: https://github.com/Starcounter/CompanyTrack/issues/264

I have built Launcher 5.0.2 with Starcounter stable 2.2.1.3234 and put that to the old warehouse. A build for 2.3.0.5406 was put to the new warehouse.

I have merged Launcher develop branch with master.

tomalec commented:
Sorry, I forgot about it :( 

Thanks :)

warpech commented:
@tomalec I will proceed with other points from https://github.com/StarcounterPrefabs/Launcher/pull/259#issuecomment-269080600

warpech commented:
Tracked in a new issue: https://github.com/StarcounterPrefabs/Launcher/issues/263

tomalec commented:
Thanks
